### ISSUE TITLE
New sprites / visual enhancements

### ISSUE THREAD

**Issue by RonRicoSuave** [DRKV333/MechTransfer#3]
Hi,

I'm noticing a minor issue where a few tiny elements don't scale correctly;  at the default zoom everything is fine.  When zoomed in with plus, they remain the same size and scroll in parallax.  It looks like the issue is MechTransferAssemblerWorld not accounting for zoom.  I know the first draw call needs to use the long form that the second does and the textures will scale properly, but the positions still mess up with simple multiplication of the x / y of the texture and I'm not sure how the screen coordinates change with scale to correct it.   On top of that calamity mod's new version is memory leaking badly (no idea how they managed that in a language that's really good at automatic garbage collection and I don't feel like extracting it again and digging through to find out) and screwing up the game after being reloaded a couple of times so I'm done with that for the night.  😸 

Anyway, this is how the bug appears.  First the ok display when zoomed out.  
![mech_ok](https://user-images.githubusercontent.com/2335114/49678889-32697d00-fa4d-11e8-8112-69d5db0a37b7.jpg)

And here's zoomed in with the items positioned and scaled incorrectly. 
![mech_zoomed](https://user-images.githubusercontent.com/2335114/49678906-4d3bf180-fa4d-11e8-9691-76a3640150cd.jpg)

Also, not related to the bug, but you might notice I started messing with the sprites.  My change to the turrets is ugly and I'm playing with different styles for the connectors but I think I'm making some improvement over the current "flat" ones, at least with the pipes / connectors.  If you're interested in those I can send them along to you for feedback once I have a consistent set.   I'm mainly shading the existing ones, but I may try to rework the turrets because they're pretty hard to flesh out as-is.

**Comment by RonRicoSuave** [comment]
...took one last look.   This fixes it for me: 

`Vector2 newPos = new Vector2(Main.GameZoomTarget * (x * 16 - Main.ViewPosition.X), 
                                                 Main.GameZoomTarget * (y * 16 - Main.ViewPosition.Y));
Main.spriteBatch.Draw(texture, newPos , null, Lighting.GetColor(x, y), 0, Vector2.Zero, Main.GameZoomTarget, SpriteEffects.None, 0);`

Similar can be done with the flipped version I'm guessing. 

Cheers

**Comment by DRKV333** [comment]
Hello,
Thanks for your report. I have actually noticed this problem on a few screenshots, but I thought not many people use the zoom feature anyway, so I just didn't bother... I've actually just found a really simple fix for this, where instead of starting my sprite batch like this:

`Main.spriteBatch.Begin(SpriteSortMode.Immediate, BlendState.AlphaBlend);`

I can just transform everything I draw by doing this:

`Main.spriteBatch.Begin(SpriteSortMode.Immediate, BlendState.AlphaBlend, Main.DefaultSamplerState, null, null, null, Main.GameViewMatrix.TransformationMatrix);`

This also fixes the assembler area rectangles. This same problem also appeared with the filter tool-tips, but I can fix that with the same transform matrix.

Those new sprites look pretty good! I really like the more rounded look on the injector. The pipes themselves might be a lost a cause at this point, so feel free to redo them as well if you feel like. Maybe something a bit thinner and more sort of steampunky would look better...

**Comment by RonRicoSuave** [comment]
Hi,

Ugh figures it was that simple.  I saw that in the main game code and missed the lack of it in the mod.  

Glad you like what's there so far.  I'm going to settle on a style and keep messing with them between playing over the next week, so I'll pass them along when I'm closer to done.  I'll see what I can come up with for the pipes.  

I hadn't played with filters or assemblers yet so cool those are working now too.  I like what I've used so far.  The only thing that felt missing  (and is in MagicStorage, too) is some kind of bridge block for walls.  Maybe just a solid block that counts as a pipe path rather than going as far as wires do with that sort of thing.  Or solid / foreground versions of the extractors / injectors might be easier to deal with.    If I get bored enough I might mess around and try to write that, doesn't seem like it should be that bad.

**Comment by DRKV333** [comment]
Well, saying that assemblers "work" would be a bit of a stretch... I've come up with a few ways to rework them, but they involve adding some fairly complicated mechanics, and I don't quite have the time for that right now. (Though actually in the near future I might...)

So, you mean there should be a solid version of the pipes? They are actually proper foreground tiles even right now, just not solid. They are 1x1 tile objects without any anchors, so you can place them anywhere, just like logic gates.

**Comment by RonRicoSuave** [comment]
Yeah, I'm thinking of either a solid version would be clearer on the purpose.  Really the only things it would be good for are keeping liquids out and enclosing NPC rooms.  **TileID.Sets.RoomNeeds.CountsAsDoor** can be set on the current ones to make them seal rooms and still be passable but it would lead to accidentally splitting rooms with existing layouts which might be a problem in somebody's setup.   Not a huge priority but it would be a useful feature.

**Comment by Flithor** [comment]
Those new sprites looks very nice!
Please be sure to use it!

**Comment by DRKV333** [comment]
I'll use them as soon as I get them. 😃

**Comment by DRKV333** [comment]
Don't want to bother you, or anything, but do you have any updates on this?

**Comment by RonRicoSuave** [comment]
Here's what I have so far;  I never did get around to fixing the turret, got distract with Real Life again.   Unfortunately the PSDs seem to have disappeared for the multi-tiles I was using them for.  These are just the .png.  I might suggest saving multi-layer .tiff rather than .xcf if Gimp supports it...  pretty much everything else does, whereas .xcf support is kinda spotty and doesn't exist in any of the higher end programs...    Corel Photo Paint is about the only thing I recognized on the compat list that's halfway  and it doesn't support all layer blending modes.  OTOH layers are only an option for the baseline TIFF standard so I don't know how widely supported they are, I'm just assuming the open source libs would implement them as well as any royalty free compression methods.  

Sorry I can't do a proper pull right now, but if I don't send these I'll forget for another 6 months.   Next time I have time to play some terraria I may work on them a bit more.  

Thanks again for the mod!    I'm a nerd and have fun building the minecraft mod style pipe item distribution systems.   Some day I'd like to implement a power / fuel system in terraria along the lines of the minecraft systems...  a 2D port of NuclearCraft with the ability to accidentally cause meltdowns if it was damaged would be kinda hilarious.  


[Tiles.zip](https://github.com/DRKV333/MechTransfer/files/3254655/Tiles.zip)

**Comment by DRKV333** [comment]
Thanks for the heads up about the .xcf files. I've actually decided I won't keep most of those around anymore. There really isn't much point. For the ones that stay, I'll convert them to .tiff instead.

So, in the mean time I have started reworking the sprites myself.

![Imgur](https://i.imgur.com/tv8MwyJ.png)

I'm trying to do them with half the resolution, like the vanilla sprites. It's kinda difficult to make all of them  easily distinguishable. I'm not sure how to deal with that. I might have to come up with some different kind of color coding for the gates and filters...

I really do like your new sprites though. Especially the turrets. I might just give up use those instead.

**Comment by DRKV333** [comment]
Well, I did some more research on those file formats. Turns out Gimp can only open multi layer tiff images, but it can't export them. It can only export each layer as an individual tiff.
Gimp can save and open simple PDS files, so that might also be an option. Wouldn't add too much extra compatibility though.
I could use Open Raster, which would be great. Photoshop doesn't support it though.
Other multi layer formats supported by Gimp are ICO and GIF...
